File: internal/executor/output_walker.go
package executor

import (
	"context"
	"fmt"

	"github.com/hsblabs/scrape-kdl/internal/ir"
)

type outputReader[Scope any] interface {
	readOutputField(Scope, ir.Field, string) (any, error)
	queryOutputRows(Scope, ir.Collection, string) ([]Scope, error)
}

type outputWalker[Scope any] struct {
	ctx        context.Context
	reader     outputReader[Scope]
	transforms *transformRuntime
	warnings   *[]Warning
	partial    *bool
}

func newOutputWalker[Scope any](ctx context.Context, reader outputReader[Scope], transforms *transformRuntime, warnings *[]Warning, partial *bool) *outputWalker[Scope] {
	return &outputWalker[Scope]{ctx: ctx, reader: reader, transforms: transforms, warnings: warnings, partial: partial}
}

func (walker *outputWalker[Scope]) executeObject(scope Scope, object ir.OutputObject, path string) (map[string]any, error) {
	result := make(map[string]any, len(object.Members))
	for _, member := range object.Members {
		if err := executionContextError(walker.ctx, path); err != nil {
			return nil, err
		}
		switch typed := member.(type) {
		case ir.Field:
			value, err := walker.executeField(scope, typed, path+"."+typed.Name)
			if err != nil {
				return nil, err
			}
			result[typed.Name] = value
		case ir.Collection:
			value, err := walker.executeCollection(scope, typed, path+"."+typed.Name)
			if err != nil {
				return nil, err
			}
			result[typed.Name] = value
		default:
			return nil, &ExecutionError{Code: "E_IR_INVALID", Message: fmt.Sprintf("unknown output member %T", member), Path: path}
		}
	}
	return result, nil
}

func (walker *outputWalker[Scope]) executeField(scope Scope, field ir.Field, path string) (any, error) {
	value, err := walker.reader.readOutputField(scope, field, path)
	if err != nil {
		if _, missing := err.(missingValue); missing {
			return handleMissingOutput(field, path, err.Error())
		}
		return recoverOutputField(walker.partial, walker.warnings, field, path, err)
	}
	value, err = walker.transforms.applyCalls(value, field.Transforms, path)
	if err != nil {
		return recoverOutputField(walker.partial, walker.warnings, field, path, err)
	}
	if !matchesRuntimeType(value, field.SuccessfulType) {
		return recoverOutputField(walker.partial, walker.warnings, field, path, &ExecutionError{Code: "E_OUTPUT_TYPE", Message: fmt.Sprintf("value of type %T is not assignable to %s", value, field.SuccessfulType.String()), Path: path})
	}
	return value, nil
}

func (walker *outputWalker[Scope]) executeCollection(scope Scope, collection ir.Collection, path string) ([]any, error) {
	rows, err := walker.reader.queryOutputRows(scope, collection, path)
	if err != nil {
		return nil, err
	}
	result := make([]any, 0, len(rows))
	for index, row := range rows {
		rowPath := fmt.Sprintf("%s[%d]", path, index)
		if err := executionContextError(walker.ctx, rowPath); err != nil {
			return nil, err
		}
		value, err := walker.executeObject(row, collection.Row, rowPath)
		if err != nil {
			if isExecutionCanceled(err) || collection.OnRowError != "skip" {
				return nil, err
			}
			*walker.partial = true
			rowIndex := index
			*walker.warnings = append(*walker.warnings, Warning{Code: "W_ROW_SKIPPED", Message: err.Error(), Path: path, Row: &rowIndex})
			continue
		}
		result = append(result, value)
		if err := validateCollectionMaximum(collection.MaxItems, len(result), path); err != nil {
			return nil, err
		}
	}
	if err := validateCollectionMinimum(collection.Required, collection.MinItems, len(result), path); err != nil {
		return nil, err
	}
	return result, nil
}

File: packages/scrape-kdl/src/output-walker.ts
import { ExecutionError } from "./execution-error.js";
import type { CollectionIR, FieldIR, JsonValue, OutputObjectIR, ResolvedTransformCallIR } from "./ir.js";
import {
  executionMessage,
  handleMissingOutput,
  MissingValue,
  type OutputState,
  recoverOutputField,
  validateCollectionMaximum,
  validateCollectionMinimum,
} from "./output-semantics.js";

export interface OutputOperations<Scope> {
  readonly state: OutputState;
  check(path: string): void;
  readField(scope: Scope, field: FieldIR, path: string): JsonValue | Promise<JsonValue>;
  queryRows(scope: Scope, collection: CollectionIR, path: string): readonly Scope[] | Promise<readonly Scope[]>;
  applyTransforms(value: JsonValue, calls: readonly ResolvedTransformCallIR[], path: string): Promise<JsonValue>;
  validateField(value: JsonValue, field: FieldIR, path: string): void;
}

export async function executeOutputObject<Scope>(
  operations: OutputOperations<Scope>,
  scope: Scope,
  object: OutputObjectIR,
  path: string,
): Promise<Record<string, JsonValue>> {
  const output: Record<string, JsonValue> = {};
  for (const member of object.members) {
    operations.check(path);
    output[member.name] =
      member.kind === "field"
        ? await executeField(operations, scope, member, `${path}.${member.name}`)
        : await executeCollection(operations, scope, member, `${path}.${member.name}`);
  }
  return output;
}

async function executeField<Scope>(
  operations: OutputOperations<Scope>,
  scope: Scope,
  field: FieldIR,
  path: string,
): Promise<JsonValue> {
  let value: JsonValue;
  try {
    value = await operations.readField(scope, field, path);
  } catch (error) {
    if (error instanceof MissingValue) return handleMissingOutput(field, path, error.message);
    return recoverOutputField(operations.state, field, path, error);
  }
  try {
    value = await operations.applyTransforms(value, field.transforms, path);
    operations.validateField(value, field, path);
    return value;
  } catch (error) {
    return recoverOutputField(operations.state, field, path, error);
  }
}

async function executeCollection<Scope>(
  operations: OutputOperations<Scope>,
  scope: Scope,
  collection: CollectionIR,
  path: string,
): Promise<readonly JsonValue[]> {
  const rows = await operations.queryRows(scope, collection, path);
  const output: JsonValue[] = [];
  for (let index = 0; index < rows.length; index++) {
    const rowPath = `${path}[${index}]`;
    operations.check(rowPath);
    try {
      output.push(await executeOutputObject(operations, rows[index]!, collection.row, rowPath));
    } catch (error) {
      if (
        (error instanceof ExecutionError && error.code === "E_EXECUTION_CANCELED") ||
        collection.onRowError !== "skip"
      )
        throw error;
      operations.state.partial = true;
      operations.state.warnings.push({ code: "W_ROW_SKIPPED", message: executionMessage(error), path, row: index });
    }
    validateCollectionMaximum(collection.maxItems, output.length, path);
  }
  validateCollectionMinimum(collection.required, collection.minItems, output.length, path);
  return output;
}
